phoenix-core/src/main/java/org/apache/phoenix/schema/json/PhoenixJson.java

```diff
@@ -0,0 +1,266 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more contributor license
+ * agreements. See the NOTICE file distributed with this work for additional information regarding
+ * copyright ownership. The ASF licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at http://www.apache.org/licenses/LICENSE-2.0 Unless required by applicable
+ * law or agreed to in writing, software distributed under the License is distributed on an "AS IS"
+ * BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License
+ * for the specific language governing permissions and limitations under the License.
+ */
+package org.apache.phoenix.schema.json;
+
+import java.io.IOException;
+import java.sql.SQLException;
+import java.util.Arrays;
+
+import org.apache.hadoop.hbase.util.Bytes;
+import org.apache.phoenix.exception.SQLExceptionCode;
+import org.apache.phoenix.exception.SQLExceptionInfo;
+import org.codehaus.jackson.JsonFactory;
+import org.codehaus.jackson.JsonNode;
+import org.codehaus.jackson.JsonParser;
+import org.codehaus.jackson.JsonParser.Feature;
+import org.codehaus.jackson.JsonProcessingException;
+import org.codehaus.jackson.map.ObjectMapper;
+import org.codehaus.jackson.node.ValueNode;
+
+import com.google.common.base.Preconditions;
+
+/**
+ * The {@link PhoenixJson} wraps json and uses Jackson library to parse and traverse the json. It
+ * should be used to represent the JSON data type and also should be used to parse Json data and
+ * read the value from it. It always conside the last value if same key exist more than once.
+ */
+public class PhoenixJson implements Comparable<PhoenixJson> {
+    private final JsonNode node;
+    private String jsonAsString;
```

Comment:
Needs suggestions to decide whether json data should be cached as byte [] or String. 

byte[]
-------

If it is byte[] following methods would not involve any String to byte [] or byte [] to String conversion 
1. toBytes(Object object, byte[] bytes, int offset)
2. toBytes(Object object)
3. toObject(byte[] bytes, int offset, int length,
			PDataType actualType, SortOrder sortOrder, Integer maxLength,
			Integer scale)


But for following, we require byte [] -> String 

1. toStringLiteral(Object o, Format formatter)

And for toObject(Object object, @SuppressWarnings("rawtypes") PDataType actualType) if actualType is Varchar we require String -> byte []


String
-----

Where as in case of String, following methods will require String -> byte []

1. toBytes(Object object, byte[] bytes, int offset)
2. toBytes(Object object)
3. toObject(byte[] bytes, int offset, int length,
			PDataType actualType, SortOrder sortOrder, Integer maxLength,
			Integer scale)

No conversion required for 

toStringLiteral(Object o, Format formatter) and toObject(Object object, @SuppressWarnings("rawtypes") PDataType actualType) if actualType is Varchar.



byte[] seems to be more efficient here but I am not sure which is most used method and will help in performance or not. 

@JamesRTaylor  @twdsilva